Starting test plan section of Windows KEP

PatrickLang opened the pull request:
I added the test plan into the outline and added the current tests that have been merged and are running in testgrid. This isn't a complete list (see the TODOs), but it's a start.

@astrieanna and @benmoss - any feedback?

Can you take a look @adelina-t and @bclau ? Do you think this format makes sense for the tests?

/sig windows

This is continuing to add content to the draft merged in #676

astrieanna commented:
@PatrickLang It seems like a very long list of tests. Is it useful to list all of them in the KEP when presumably tagging in the code will be the source of truth on that?

PatrickLang commented:
@astrieanna yeah, I know the list is long but I was getting questions and feedback on "does X work" last time. Let me see if I can get someone from SIG-Testing to clarify what level of detail they'd like to see here.

PatrickLang commented:
/milestone v1.14

benmoss commented:
@PatrickLang do we have any plan for keeping track of & making progress on non-conformance tests? We were just looking at the tests for port-forwarding, and with our change some of them pass, and the images were already part of the set ported to Windows, but they aren't being run in Testgrid right now.

jdumars commented:
/hold

jdumars commented:
/lgtm
SIG stakeholders, feel free to remove hold at your discretion.

spiffxp commented:
So, I'm torn since I'm coming from multiple perspectives.

A) This does feel like a lot to enumerate and I’m not sure it’s sig-testing’s job to validate the exhaustive enumeration, but I suspect we have a say in whether the plan seems reasonable.  It does seem reasonable to me.

B) As a consumer of this information from the Release Team, I think I'm just asking for a dashboard, or list of dashboards to check.  So I would be fine if your test plan was just: "we're going to call this done when all tests at [sig-windows/the-blocking-tests] are green", or "[sig-windows/existing-tests], [sig-windows/substitute-tests], [sig-windows/windows-only-tests]" or something.

To me that then begs the question of whether getting to green by PR'ing failing tests out of those dashboards is OK, or whether the list of tests in those dashboards is OK.

C) As someone who watched SIG Arch flail during 1.13, I feel like there is a desire for exhaustive enumeration of some sort, to figure out whether these tests, or whether the grouping of tests is sufficient to meet user expectations for "how a kubernetes behaves".  Reviewing that is something that, to me, seems like it's in SIG Arch's purview (I am super open to suggestions on which SIGs should be involved/federated to instead)

In the interest of an initial pass at this, I'd be in favor of A) & B) where you just describe the test plan and dashboards.  I would run by sig-arch whether they need more, or would be comfortable using the dashboards and evaluating at some checkpoint.

PatrickLang commented:
Thanks @spiffxp . I'll change it to a summary level of detail now for A/B, and if other reviewers want to review it test case by test case - git to the rescue!

mattfarina commented:
Something I'm considering is perspective of those looking in. When we reviewed the tests in the SIG Arch conversation on Windows GA we looked at a dashboard of failing tests. I would suggest having a dashboard of windows tests and that they be generally as green as those for Linux nodes.

I like the plan of updating tests cases to handle multi-os, having windows specific tests, and having substitute tests cases. With my SIG Apps hat on, seeing the workloads APIs passing on Windows would be fantastic.

michmike commented on keps/sig-windows/0000-20190103-windows-node-support.md:
```diff
@@ -2,8 +2,9 @@
+  - "@patricklang"
 owning-sig: sig-windows
 participating-sigs:
   - sig-architecture
```
as per the discussion today, sig-testing and sig-release should be reviewers for this part of the KEP

michmike commented on keps/sig-windows/0000-20190103-windows-node-support.md:
```diff
@@ -106,6 +107,196 @@ As of 29-11-2018 much of the work for enabling Windows nodes has already been co
+1. [Adapting](#Adapting-existing-tests) some of the existing conformance tests to be able to pass on multiple node OS's
+2. Adding [substitute](#Substitute-test-cases) test cases where the first approach isn't feasible or would change the tests in a way is not approved by the owner. These will be tagged with `[SIG-Windows]`
+3. Last, gaps will be filled with [Windows specific tests](#Windows-specific-tests). These will also be tagged with `[SIG-Windows]`
+
```
I think it is important to note that any test that is tagged with `[SIG-Windows]` should always pass on windows. we should specify this in the text.

In addition, is it possible to tag any of the tests that are not applicable to windows with any tag? that way it is easy to differentiate which of the existing conformance tests are not applicable to windows (beyond the ones that were substituted)

michmike commented:
/approve
/lgtm

michmike commented:
> So, I'm torn since I'm coming from multiple perspectives.
> 
> A) This does feel like a lot to enumerate and I’m not sure it’s sig-testing’s job to validate the exhaustive enumeration, but I suspect we have a say in whether the plan seems reasonable. It does seem reasonable to me.
> 
> B) As a consumer of this information from the Release Team, I think I'm just asking for a dashboard, or list of dashboards to check. So I would be fine if your test plan was just: "we're going to call this done when all tests at [sig-windows/the-blocking-tests] are green", or "[sig-windows/existing-tests], [sig-windows/substitute-tests], [sig-windows/windows-only-tests]" or something.
> 
> To me that then begs the question of whether getting to green by PR'ing failing tests out of those dashboards is OK, or whether the list of tests in those dashboards is OK.
> 
> C) As someone who watched SIG Arch flail during 1.13, I feel like there is a desire for exhaustive enumeration of some sort, to figure out whether these tests, or whether the grouping of tests is sufficient to meet user expectations for "how a kubernetes behaves". Reviewing that is something that, to me, seems like it's in SIG Arch's purview (I am super open to suggestions on which SIGs should be involved/federated to instead)
> 
> In the interest of an initial pass at this, I'd be in favor of A) & B) where you just describe the test plan and dashboards. I would run by sig-arch whether they need more, or would be comfortable using the dashboards and evaluating at some checkpoint.

i think for this initial iteration of the KEP, we do need (a), (b), and (c) to satisfy curious minds. Ultimately the dashboard with green checkboxes across all tests applicable to windows is what most folks will be looking for, but the approvers of the KEP will likely need to dig in.

bgrant0607 commented:
cc @BenTheElder

BenTheElder commented on keps/sig-windows/0000-20190103-windows-node-support.md:
```diff
@@ -106,6 +107,196 @@ As of 29-11-2018 much of the work for enabling Windows nodes has already been co
+
+> TODO: verify against list of test cases from https://docs.google.com/document/d/1YkLZIYYLMQhxdI2esN5PuTkhQHhO0joNvnbHpW68yg8/edit#
+
+- [k8s.io] Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart exec hook properly [NodeConformance] [Conformance]
```
Seems like SIG-Arch should be OKaying this, as it's sort of just a round-about way of having a [Windows] profile?

BenTheElder replied:
(Totally understand the intent here, just not sure it's the right place to review and/or if that review already occurred).

BenTheElder commented on keps/sig-windows/.gitignore:
```diff
@@ -0,0 +1,2 @@
+*.swp
+*~
```
seems like probably this should be in the root .gitignore? :-)

PatrickLang replied:
moved into https://github.com/kubernetes/enhancements/pull/702

PatrickLang commented:
Updated areas for reviewers:
- removed .gitignore, put into #702 
- Added more details including links to untested PRs in test approaches
- Added more detail to features not supported that we listed in the [previous doc](https://docs.google.com/document/d/1YkLZIYYLMQhxdI2esN5PuTkhQHhO0joNvnbHpW68yg8/edit#)

Another PR will include more API specifics. There's two things we'll do:
- Review the Pod, Container, and SecurityContext line by line
- (this will be late Jan) - I'm checking for API coverage data from the existing tests to find more gaps. I don't know 100% if this will work or not since I just started investigating it.

PatrickLang commented:
@justaugustus removed KEP number & renamed

bclau commented on keps/sig-windows/20190103-windows-node-support.md:
```diff
@@ -0,0 +1,195 @@
+
+- [ ] [sig-network] [sig-windows] Networking Granular Checks: Pods should function for intra-pod communication: http - [PR#71468](https://github.com/kubernetes/kubernetes/pull/71468)
+- [ ] [sig-network] [sig-windows] Networking Granular Checks: Pods should function for intra-pod communication: udp - [PR#71468](https://github.com/kubernetes/kubernetes/pull/71468)
+- [ ] [sig-network] [sig-windows] Networking Granular Checks: Pods should function for node-pod communication: udp - [PR#71468](https://github.com/kubernetes/kubernetes/pull/71468)
```
For the sake of completeness, you should also include: ``[sig-network] [sig-windows] Networking Granular Checks: Pods should function for node-pod communication: http``, same PR.

PatrickLang replied:
Oops my mistake. Adding now

bclau commented:
As far the tests and the required patches on the testing side, other than the comment above, LGTM.

adelina-t commented:
LGTM

michmike commented:
/lgtm
/approve

michmike commented:
/hold cancel

michmike commented on keps/sig-windows/20190103-windows-node-support.md:
```diff
@@ -0,0 +1,196 @@
+
+### Non-Goals
+
+- Adding Windows support to all projects in the Kubernetes ecosystem (Cluster Lifecycle, etc)
```
PR 729 already had this change as `Enable the Kubernetes master components to run on Windows`

bgrant0607 commented on keps/sig-windows/20190103-windows-node-support.md:
```diff
@@ -0,0 +1,196 @@
+Additional dashboard pages will be added over time as we run the same test cases with additional CRI, CNI and cloud providers. They reflect work that may be stabilized in v1.15 or later and is not strictly required for v1.14.
+
+- Windows Server 2019 on GCP - this is [in progress](https://k8s-testgrid.appspot.com/google-windows#windows-prototype), but not required for v1.14
+- Windows Server 2019 with OVN+OVS & Dockershim
```
Are these additional dashboards? Are they all required other than GCP? What configuration will be shown on the main dashboard?

PatrickLang replied:
Will clarify in the next update on #732

bgrant0607 commented on keps/sig-windows/20190103-windows-node-support.md:
```diff
@@ -0,0 +1,196 @@
+
+The testing for Windows nodes will include multiple approaches:
+
+1. [Adapting](#Adapting-existing-tests) some of the existing conformance tests to be able to pass on multiple node OS's. Tests that won't work will be [excluded](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/sig-windows/sig-windows-config.yaml#L69).
```
I believe those tests are all being tagged [LinuxOnly].
https://github.com/kubernetes/kubernetes/pull/73204
https://github.com/kubernetes/community/pull/3140

benmoss commented on keps/sig-windows/20190103-windows-node-support.md:
```diff
@@ -0,0 +1,196 @@
+
+All of the test cases will be maintained within the kubernetes/kubernetes repo. SIG-Windows specific tests for 2/3 will be in [test/e2e/windows](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/windows)
+
+Additional Windows test setup scripts, container image source code, and documentation will be kept in the [kubernetes-sigs/windows-testing](https://github.com/kubernetes-sigs/windows-testing) repo. One example is that the prow jobs need a list of repos to use for the test containers, and that will be maintained here - see [windows-testing#1](https://github.com/kubernetes-sigs/windows-testing/issu ...
```
Right now they're in https://hub.docker.com/u/e2eteam but ideally we can consolidate these in the same registry as the other Kubernetes images

PatrickLang replied:
These could be pushed to GCR if someone gives us the right access.

We can't build Windows containers in the main build today, but that's the eventual goal. Building these at the same cadence will require adding Windows nodes into the build pool, or moving all images to languages that can be cross-compiled and put into containers with only offline operations (no RUN commands in Dockerfile).
